src/clj/rems/auth/oidc.clj

```diff
@@ -23,25 +27,89 @@
-  (with-redefs [env {:oidc-userid-attribute ["atr" "sub" "fallback"]}]
+  (with-redefs [env {:oidc-userid-attributes [{:attribute "atr"}
+                                              {:attribute "sub"}
+                                              {:attribute "fallback"}]}]
     (is (= "123" (get-userid {:sub "123"})))
     (is (= "456" (get-userid {:sub "123" :atr "456"})))
     (is (= "456" (get-userid {:sub "123" :atr "456" :fallback "78"})))
-    (is (= "78" (get-userid {:fallback "78"})))))
+    (is (= "78" (get-userid {:fallback "78"}))))
+  (with-redefs [env {:oidc-userid-attributes [{:attribute "atr" :rename "elixirId"}
+                                              {:attribute "sub"}]}]
+    (is (= "elixir-alice" (get-userid {:elixirId "elixir-alice" :sub "123"})))))
+
+(defn rename-userid-attributes [id-data]
+  (let [attrs (getx env :oidc-userid-attributes)]
+    (->> attrs
+         (filter :rename)
+         (map (partial map-vals keyword))
+         (reduce #(replace-key %1 (:attribute %2) (:rename %2)) id-data))))
+
+(deftest test-rename-user-attributes
+  (with-redefs [env {:oidc-userid-attributes [{:attribute "sub" :rename "elixirId"}
+                                              {:attribute "old_sub"}]}]
+    (is (nil? (rename-userid-attributes nil)))
+    (is (= #{[:old_sub "alice"]
+             [:elixirId "elixir-alice"]
+             [:name "Alice Applicant"]}
+           (set (rename-userid-attributes {:old_sub "alice"
+                                           :sub "elixir-alice"
+                                           :name "Alice Applicant"}))))))
+
+(defn find-or-create-user! [id-data]
+  (let [userid-attrs (get-userid-attributes id-data)]
+    (or (some #(user-mappings/get-user-mapping (first %) (second %)) userid-attrs)
+        (find-first users/user-exists? (vals userid-attrs))
+        (let [userid (get-userid id-data)]
+          (users/add-user-raw! userid id-data)
+          userid))))
+
+(defn save-user-mappings! [id-data userid]
+  (let [attrs (->> (getx env :oidc-userid-attributes)
+                   (map #(or (:rename %) (:attribute %)))
+                   (map keyword))]
+    (doseq [[attr value] (select-keys id-data attrs)
+            :when (not= value userid)]
+      (user-mappings/create-user-mapping! {:userid userid
+                                           :ext-id-attribute (name attr)
+                                           :ext-id-value value}))))
+
+(defn get-user-attributes [id-data]
+  (let [identity-base {:eppn (get-userid id-data)
+                       ;; need to maintain a fallback list of name attributes since identity
+                       ;; providers differ in what they give us
+                       :commonName (some id-data [:name :unique_name :family_name])
+                       :mail (:email id-data)}
+        extra-attributes (select-keys id-data (map (comp keyword :attribute) (:oidc-extra-attributes env)))]
```

Comment:
This could in theory support rename, and the email etc. could be handled through something? We can pull these out as an improvement ticket.